File: psy_now/lib/services/remote_manifest.dart
import 'dart:io';

import 'package:fast_log/fast_log.dart';

/// Same contract as SalsaNOW `RemoteManifest`: [envManifestBase] or
/// `SalsaNOW.manifest.ini` next to the executable with `ManifestBaseUrl=...`.
class RemoteManifest {
  RemoteManifest._();

  static const envManifestBase = 'SALSANOW_MANIFEST_BASE';
  static const iniFileName = 'SalsaNOW.manifest.ini';

  static String? _baseUrl;

  static bool get isInitialized =>
      _baseUrl != null && _baseUrl!.trim().isNotEmpty;

  /// Returns false if neither env nor ini provides a base (caller may use a local fallback path only).
  static Future<bool> initialize() async {
    _baseUrl = Platform.environment[envManifestBase]?.trim();
    _baseUrl = _trimTrailingSlash(_baseUrl);

    if (_baseUrl == null || _baseUrl!.isEmpty) {
      _baseUrl = await _readManifestBaseFromIni();
    }

    if (_baseUrl == null || _baseUrl!.isEmpty) {
      warn(
        '[RemoteManifest] No manifest host. Set $envManifestBase or place $iniFileName beside the executable with ManifestBaseUrl=https://...',
      );
      return false;
    }

    info('[RemoteManifest] Base URL: $_baseUrl');
    return true;
  }

  /// Throws if not initialized.
  static String url(String relativePath) {
    if (!isInitialized) {
      throw StateError(
        'RemoteManifest is not configured. Set $envManifestBase or $iniFileName.',
      );
    }
    final rel = relativePath.replaceFirst(RegExp(r'^/+'), '');
    return '$_baseUrl/$rel';
  }

  static String? _trimTrailingSlash(String? s) {
    if (s == null) return null;
    var t = s.trim();
    while (t.endsWith('/')) {
      t = t.substring(0, t.length - 1);
    }
    return t.isEmpty ? null : t;
  }

  static Future<String?> _readManifestBaseFromIni() async {
    try {
      final exeDir = File(Platform.resolvedExecutable).parent.path;
      final path = '$exeDir${Platform.pathSeparator}$iniFileName';
      final f = File(path);
      if (!await f.exists()) return null;

      for (final raw in await f.readAsLines()) {
        var line = raw.trim();
        if (line.isEmpty || line.startsWith('#') || line.startsWith(';')) {
          continue;
        }
        final eq = line.indexOf('=');
        if (eq <= 0) continue;
        final key = line.substring(0, eq).trim();
        if (!key.toLowerCase().contains('manifestbaseurl')) continue;
        return _trimTrailingSlash(line.substring(eq + 1).trim());
      }
    } catch (e) {
      warn('[RemoteManifest] Error reading ini: $e');
    }
    return null;
  }

  /// For tests.
  static void resetForTests() => _baseUrl = null;
}

File: SalsaNOW/RemoteManifest.cs
using System;
using System.IO;

namespace SalsaNOW
{
    /// <summary>
    /// Resolves the root URL for bundled JSON/config hosted alongside the app (no host baked into source).
    /// Set <c>SALSANOW_MANIFEST_BASE</c> or place <c>SalsaNOW.manifest.ini</c> next to the executable with
    /// <c>ManifestBaseUrl=https://your-host</c> (no trailing slash required).
    /// </summary>
    internal static class RemoteManifest
    {
        private const string EnvManifestBase = "SALSANOW_MANIFEST_BASE";
        private const string IniFileName = "SalsaNOW.manifest.ini";
        private static string _baseUrl;

        internal static void ResetForTests() => _baseUrl = null;

        public static void Initialize()
        {
            _baseUrl = Environment.GetEnvironmentVariable(EnvManifestBase)?.Trim().TrimEnd('/');
            if (string.IsNullOrEmpty(_baseUrl))
                _baseUrl = ReadManifestBaseFromIni();

            if (string.IsNullOrEmpty(_baseUrl))
            {
                throw new InvalidOperationException(
                    "Remote manifest host is not configured. Set environment variable " + EnvManifestBase +
                    " to your base URL (example: https://cdn.example.com), or create \"" + IniFileName +
                    "\" in the application folder with a line: ManifestBaseUrl=https://cdn.example.com");
            }
        }

        public static string Url(string relativePath)
        {
            if (string.IsNullOrEmpty(_baseUrl))
                throw new InvalidOperationException("RemoteManifest.Initialize was not called.");
            if (string.IsNullOrEmpty(relativePath))
                throw new ArgumentException("Relative path is required.", nameof(relativePath));

            relativePath = relativePath.TrimStart('/');
            return $"{_baseUrl}/{relativePath}";
        }

        private static string ReadManifestBaseFromIni()
        {
            string path = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, IniFileName);
            if (!File.Exists(path))
                return null;

            foreach (var raw in File.ReadAllLines(path))
            {
                var line = raw.Trim();
                if (line.Length == 0 || line[0] == '#' || line[0] == ';')
                    continue;

                int eq = line.IndexOf('=');
                if (eq <= 0)
                    continue;

                string key = line.Substring(0, eq).Trim();
                if (!key.Equals("ManifestBaseUrl", StringComparison.OrdinalIgnoreCase))
                    continue;

                return line.Substring(eq + 1).Trim().TrimEnd('/');
            }

            return null;
        }
    }
}
